cocotb_test/simulator.py

```diff
@@ -491,40 +491,20 @@ def get_define_commands(self, defines):
     def get_parameter_commands(self, parameters):
         return ["-g" + name + "=" + str(value) for name, value in parameters.items()]
 
-    def build_script_sim(self):
-        if self.toplevel_lang == "vhdl":
-            sim_script = "vsim -onfinish {ONFINISH} -foreign {EXT_NAME} {SIM_ARGS} {EXTRA_ARGS} {RTL_LIBRARY}.{TOPLEVEL};".format(
-                ONFINISH="stop" if self.gui else "exit",
-                RTL_LIBRARY=as_tcl_value(self.toplevel_library),
-                TOPLEVEL=as_tcl_value(self.toplevel_module),
-                EXT_NAME=as_tcl_value("cocotb_init {}".format(cocotb.config.lib_name_path("fli", "questa"))),
-                SIM_ARGS=" ".join(self.simulation_args),
-                EXTRA_ARGS=" ".join(as_tcl_value(v) for v in self.get_parameter_commands(self.parameters)),
-            )
-        else:
-            sim_script = "vsim -onfinish {ONFINISH} -pli {EXT_NAME} {SIM_ARGS} {EXTRA_ARGS} {RTL_LIBRARY}.{TOPLEVEL} {PLUS_ARGS};".format(
-                ONFINISH="stop" if self.gui else "exit",
-                RTL_LIBRARY=as_tcl_value(self.toplevel_library),
-                TOPLEVEL=as_tcl_value(self.toplevel_module),
-                EXT_NAME=as_tcl_value(cocotb.config.lib_name_path("vpi", "questa")),
-                SIM_ARGS=" ".join(self.simulation_args),
-                EXTRA_ARGS=" ".join(as_tcl_value(v) for v in self.get_parameter_commands(self.parameters)),
-                PLUS_ARGS=" ".join(as_tcl_value(v) for v in self.plus_args),
-            )
-
+    def do_script(self):
+        do_script = ""
         if self.waves:
-            sim_script += "log -recursive /*;"
-
-        return sim_script
-
-    def build_script_run(self):
-
-        return "run -all; quit"
+            do_script += "log -recursive /*;"
+        return do_script
 
     def build_command(self):
 
         cmd = []
 
+        do_script = self.do_script()
+        if not self.gui:
+            do_script += "run -all; quit"
```

Comment:
Would add this also to `do_script`.  Can imagine someone wants to stop in between and execute some command (like save/restore simulation state)

Opinion?

Reply:
True, users then can prepend their custom scripts like this
```python
class QuestaCustom(Questa):
    def do_script(self):
        do_script = f"coverage save -onexit {self.module}.ucdb;"
        do_script += super().do_script()
        return do_script
```

Reply:
👍🏻